producer/writer/message_writer.go

```diff
@@ -0,0 +1,378 @@
+
+func newMessageWriterMetrics(scope tally.Scope) messageWriterMetrics {
+	return messageWriterMetrics{
+		consumerWriteError: scope.Counter("consumer-write-error"),
+		writeError:         scope.Counter("write-error"),
+		writeSuccess:       scope.Counter("write-success"),
+		writeRetry: scope.
+			Tagged(map[string]string{"write-type": "retry"}).
+			Counter("write"),
+		writeNew: scope.
+			Tagged(map[string]string{"write-type": "new"}).
+			Counter("write"),
+		writeAfterCutoff: scope.
+			Tagged(map[string]string{"reason": "after-cutoff"}).
+			Counter("invalid-write"),
+		writeBeforeCutover: scope.
+			Tagged(map[string]string{"reason": "before-cutover"}).
+			Counter("invalid-write"),
+	}
+}
+
+type messageWriterImpl struct {
+	sync.RWMutex
+
+	replicatedShardID uint64
+	mPool             messagePool
+	opts              Options
+
+	msgID           uint64
+	queue           *list.List
+	consumerWriters map[string]consumerWriter
+	acks            *acks
+	cutOffNanos     int64
+	cutOverNanos    int64
+	isClosed        bool
+	doneCh          chan struct{}
+	wg              sync.WaitGroup
+	m               messageWriterMetrics
+
+	nowFn clock.NowFn
+}
+
+func newMessageWriter(
+	replicatedShardID uint64,
+	mPool messagePool,
+	opts Options,
+) messageWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	return &messageWriterImpl{
+		replicatedShardID: replicatedShardID,
+		mPool:             mPool,
+		opts:              opts,
+		msgID:             0,
+		queue:             list.New(),
+		consumerWriters:   make(map[string]consumerWriter),
+		acks:              newAckHelper(defaultAckMapSize),
+		cutOffNanos:       0,
+		cutOverNanos:      0,
+		isClosed:          false,
+		doneCh:            make(chan struct{}),
+		m:                 newMessageWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:             time.Now,
+	}
+}
+
+func (w *messageWriterImpl) Write(rd producer.RefCountedData) {
+	nowNanos := w.nowFn().UnixNano()
+	w.RLock()
+	isValid := w.isValidWriteWithLock(nowNanos)
+	w.RUnlock()
+	if !isValid {
+		return
+	}
+	rd.IncRef()
+	msg := w.mPool.Get()
+
+	w.Lock()
+	w.msgID++
+	meta := metadata{
+		shard: w.replicatedShardID,
+		id:    w.msgID,
+	}
+	msg.Reset(meta, rd)
+	w.acks.add(meta, msg)
+	w.writeWithLock(msg, nowNanos)
+	w.queue.PushBack(msg)
+	w.Unlock()
+
+	w.m.writeNew.Inc(1)
+}
+
+func (w *messageWriterImpl) isValidWriteWithLock(nowNanos int64) bool {
+	if w.cutOffNanos > 0 && nowNanos >= w.cutOffNanos {
+		w.m.writeAfterCutoff.Inc(1)
+		return false
+	}
+	if w.cutOverNanos > 0 && nowNanos < w.cutOverNanos {
+		w.m.writeBeforeCutover.Inc(1)
+		return false
+	}
+	return true
+}
+
+func (w *messageWriterImpl) writeWithLock(m *message, nowNanos int64) {
+	m.IncReads()
+	msg, isValid := m.Marshaler()
+	if !isValid {
+		m.DecReads()
+		return
+	}
+	written := false
+	for _, cw := range w.consumerWriters {
+		if err := cw.Write(msg); err != nil {
+			w.m.consumerWriteError.Inc(1)
+			continue
+		}
+		written = true
```

Comment:
Hm so we only write the message to one writer? Is this a shared writer?

Reply:
see comment above

Reply:
I'm a bit confused, if a `shardWriter` can have multiple `messageWriter`s, does that mean each `messageWriter` should correspond to a consumer instance then? What's the relationship between `messageWriter` and `consumerWriter`?

Reply:
a messageWriter could have one or multiple consumerWriters, depend on the consumption type. 
In shared mode, the shardWriter will only maintain one messageWriter and can add multiple consumerWriters to the messageWriter.
In replicated mode, the shardWriter will maintain one messageWriter for each replica and only keep one consumerWriter per messageWriter